blog/2024-02-01-py-client-v4-release/index.mdx

```diff
@@ -0,0 +1,345 @@
+
+<FilteredTextBlock
+    text={PythonCode}
+    startMarker="# TryFinallyExample"
+    endMarker="# END TryFinallyExample"
+    language="py"
+/>
+
+### Batch operations
+
+The new client provides a set of methods to configure batching. These methods help you to control the batching behavior - including a handy `rate_limit()` method to avoid going past any third-party API rate limits.
+
+| Method | Description | When to use |
+| :-- | :-- | :-- |
+| `dynamic` | The batch size is dynamically calculated by Weaviate. | Recommended starting point. |
+| `fixed_size` | The batch size is fixed to a size specified by a user. | When you want to control the batch size. |
+| `rate_limit` | The number of objects sent to Weaviate is rate limited (specified as n_objects per minute). | When you want to avoid hitting third-party vectorization API rate limits. |
+
+<Tabs groupId="languages">
+    <TabItem value="wcs" label="Dynamic">
+        <FilteredTextBlock
+            text={PythonCode}
+            startMarker="# START BatchDynamic"
+            endMarker="# END BatchDynamic"
+            language="py"
+        />
+    </TabItem>
+    <TabItem value="local" label="Fixed size">
+        <FilteredTextBlock
+            text={PythonCode}
+            startMarker="# START BatchFixedSize"
+            endMarker="# END BatchFixedSize"
+            language="py"
+        />
+    </TabItem>
+    <TabItem value="embedded" label="Rate limit">
+        <FilteredTextBlock
+            text={PythonCode}
+            startMarker="# START BatchRateLimit"
+            endMarker="# END BatchRateLimit"
+            language="py"
+        />
+    </TabItem>
+</Tabs>
+
+There's also single insertion methods, of course. And to that, we've added a `insert_many` method which is a convenience method for inserting a list of objects.
+
+<Tabs groupId="languages">
+    <TabItem value="batching" label="Batch">
+        <FilteredTextBlock
+            text={PythonCode}
+            startMarker="# START BatchBasic"
+            endMarker="# END BatchBasic"
+            language="py"
+        />
+        <i>Generally recommended for any multi-object operations.</i>
+    </TabItem>
+    <TabItem value="insert_many" label="Insert many">
+        <FilteredTextBlock
+            text={PythonCode}
+            startMarker="# START InsertManyBasic"
+            endMarker="# END InsertManyBasic"
+            language="py"
+        />
+        <i>If you just have a list of objects to insert.</i>
+    </TabItem>
+    <TabItem value="insert" label="Single insert">
+        <FilteredTextBlock
+            text={PythonCode}
+            startMarker="# START CreateObjectExample"
+            endMarker="# END CreateObjectExample"
+            language="py"
+        />
+        <i>For single-insertions; not for bulk operations.</i>
+    </TabItem>
+</Tabs>
```

Comment:
a short example for error handling? :) That was a pain in v3 because you got a print into the console or had to provide some function.

I think there are two things to know:
- breaking a loop if there are too many errors:
```
with client.batch.dynamic() as batch:
        if  batch.number_errors > 10:
            break
```

and checking afterwards the list of errors
```
client.batch.failed_objects  # also refs
```

Reply:
Added.